common/network-common/src/main/java/org/apache/spark/network/sasl/aes/SparkAesCipher.java

```diff
@@ -0,0 +1,270 @@
+
+/**
+ * AES cipher for encryption and decryption.
+ */
+public class SparkAesCipher {
+  private static final Logger logger = LoggerFactory.getLogger(SparkAesCipher.class);
+  private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+  public final static String supportedTransformation[] = {
+    "AES/CBC/NoPadding", "AES/CTR/NoPadding"
+  };
+
+  private final CryptoCipher encryptor;
+  private final CryptoCipher decryptor;
+
+  private final Integrity integrity;
+
+  public SparkAesCipher(
+      String cipherTransformation,
+      Properties properties,
+      byte[] inKey,
+      byte[] outKey,
+      byte[] inIv,
+      byte[] outIv) throws IOException {
+    if (!Arrays.asList(supportedTransformation).contains(cipherTransformation)) {
+      logger.warn("AES cipher transformation is not supported: " + cipherTransformation);
+      cipherTransformation = "AES/CTR/NoPadding";
+      logger.warn("Use default AES/CTR/NoPadding");
+    }
+
+    final SecretKeySpec inKeySpec = new SecretKeySpec(inKey, "AES");
+    final IvParameterSpec inIvSpec = new IvParameterSpec(inIv);
+    final SecretKeySpec outKeySpec = new SecretKeySpec(outKey, "AES");
+    final IvParameterSpec outIvSpec = new IvParameterSpec(outIv);
+
+    // Encryptor
+    encryptor = Utils.getCipherInstance(cipherTransformation, properties);
+    try {
+      logger.debug("Initialize encryptor");
+      encryptor.init(Cipher.ENCRYPT_MODE, outKeySpec, outIvSpec);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to initialize encryptor", e);
+    }
+
+    // Decryptor
+    decryptor = Utils.getCipherInstance(cipherTransformation, properties);
+    try {
+      logger.debug("Initialize decryptor");
+      decryptor.init(Cipher.DECRYPT_MODE, inKeySpec, inIvSpec);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to initialize decryptor", e);
+    }
+
+    integrity = new Integrity(outKey, inKey);
+  }
+
+  /**
+   * Encrypts input data. The result composes of (msg, padding if needed, mac) and sequence num.
+   * @param data the input byte array
+   * @param offset the offset in input where the input starts
+   * @param len the input length
+   * @return the new encrypted byte array.
+   * @throws SaslException if error happens
+   */
+  public byte[] wrap(byte[] data, int offset, int len) throws SaslException {
+    byte[] mac = integrity.getHMAC(data, offset, len);
+    integrity.incMySeqNum();
+
+    // Padding based on cipher
+    byte[] padding;
+    if ("AES/CBC/NoPadding".equals(encryptor.getAlgorithm())) {
+      int bs = encryptor.getBlockSize();
+      int pad = bs - (len + 10) % bs;
+      padding = new byte[pad];
+      for (int i = 0; i < pad; i ++) {
+        padding[i] = (byte) pad;
+      }
+    } else {
+      padding = EMPTY_BYTE_ARRAY;
+    }
+
+    // Encrypt
+    byte[] encrypted = new byte[len + 10 + padding.length + 4];
```

Comment:
Aside from the "why so much copying" and "why not use ByteBuffer" comments, there's a lot of magic numbers being used around this code...